File: 54-spiral-matrix/spiral-matrix.cpp
class Solution {
public:
    vector<int> spiralOrder(vector<vector<int>>& matrix) {
        vector<int> res;
        int ROWS = matrix.size();
        int COLS = matrix[0].size();
        int t = 0, b = ROWS - 1;
        int l = 0, r = COLS - 1;
        string direction = "right";
        while(l <= r && t <= b) {
            if(direction == "right") {
                for(int i=l; i<=r; i++) {
                    res.push_back(matrix[t][i]);
                }
                t++;
                direction = "down";
            }
            else if(direction == "down") {
                for(int i=t; i<=b; i++) {
                    res.push_back(matrix[i][r]);
                }
                r--;
                direction = "left";
            }
            else if(direction == "left") {
                for(int i=r; i >= l; i--) {
                    res.push_back(matrix[b][i]);
                }
                b--;
                direction = "top";
            }
            else if(direction == "top") {
                for(int i=b; i >= t; i--) {
                    res.push_back(matrix[i][l]);
                }
                l++;
                direction = "right";
            }
        }  
        return res;
    }
};

File: 54-spiral-matrix/spiral-matrix.py
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        l, r, t, b = 0, len(matrix[0]), 0, len(matrix)
        ROWS, COLS = b, r
        res = []
        L, R, U, D = 'L', 'R', 'U', 'D'
        d = R
        i, j = 0, 0
        while t < b and l < r:
            res.append(matrix[i][j])
            if d == R: 
                if j == r - 1:
                    t += 1
                    d = D
                    i += 1
                    continue
                j += 1
            
            if d == D: 
                if i == b - 1:
                    r -= 1
                    d = L
                    j -= 1
                    continue
                i += 1
            
            if d == L:
                if j == l:
                    b -= 1
                    d = U
                    i -= 1
                    continue
                j -= 1
            
            if d == U:
                if i == t:
                    l += 1
                    d = R
                    j += 1
                    continue
                i -= 1

        return res
